lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
 }
 
 // QUICServer is a proxy peering server that uses the QUIC protocol.
-type QUICServer struct{}
+type QUICServer struct {
+	log           *slog.Logger
+	clusterDialer ClusterDialer
+	tlsConfig     *tls.Config
+	quicConfig    *quic.Config
+
+	mu     sync.Mutex
+	closed bool
+	wg     sync.WaitGroup
+
+	replayStore replayStore
+
+	// runCtx is a context that gets canceled when all connections should be
+	// ungracefully terminated.
+	runCtx context.Context
+	// runCancel cancels runCtx.
+	runCancel context.CancelFunc
+	// serveCtx is a context that gets canceled when all listeners should stop
+	// accepting new connections.
+	serveCtx context.Context
+	// serveCancel cancels serveCtx.
+	serveCancel context.CancelFunc
+}
 
 // NewQUICServer returns a [QUICServer] with the given config.
 func NewQUICServer(cfg QUICServerConfig) (*QUICServer, error) {
 	if err := cfg.checkAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	panic("QUIC proxy peering is not implemented")
+
+	tlsConfig := utils.TLSConfig(cfg.CipherSuites)
+	tlsConfig.GetCertificate = cfg.GetCertificate
+	tlsConfig.VerifyPeerCertificate = verifyPeerCertificateIsProxy
+	tlsConfig.NextProtos = []string{quicNextProto}
+	tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	tlsConfig.MinVersion = tls.VersionTLS13
+
+	getClientCAs := cfg.GetClientCAs
+	tlsConfig.GetConfigForClient = func(chi *tls.ClientHelloInfo) (*tls.Config, error) {
+		clientCAs, err := getClientCAs(chi)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		utils.RefreshTLSConfigTickets(tlsConfig)
+		c := tlsConfig.Clone()
+		c.ClientCAs = clientCAs
+		return c, nil
+	}
+
+	quicConfig := &quic.Config{
+		MaxStreamReceiveWindow:     quicMaxReceiveWindow,
+		MaxConnectionReceiveWindow: quicMaxReceiveWindow,
+
+		MaxIncomingStreams:    1 << 60,
```

Comment:
Is there a specific reason for `1<<60` or is it just a large number? Add a comment?

```suggestion
		MaxIncomingStreams:    1 << 60, // arbitrary, large number
```

Reply:
It's a quic-go magic number with no matching const, the maximum possible value for `MaxIncomingStreams` 😢

I'll add a comment pointing that out.

Reply:
const'd in 5fad0d54d9